### PR TITLE
issue #1929 - search optimization for select token queries without a system

### PR DESCRIPTION
**Pull request by lmsurpre**
1. Add a utility under fhir-search for adding an implicit-sytem extension to applicable search parameters

2. Use the extensions that we put in the SearchParameter definitions to set
the system of the QueryParameterValue object

Signed-off-by: Lee Surprenant <lmsurpre@us.ibm.com>

### CODE REVIEW

File: fhir-search/src/test/java/com/ibm/fhir/search/util/SearchParameterAugmenter.java [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+
+/**
+ * A utility for adding the http://ibm.com/fhir/extension/implicit-system extension
+ * to search parameters which always reference code values with a particular system
+ */
+public class SearchParameterAugmenter {
+    private static final String IMPLICIT_SYSTEM_EXT_URL = "http://ibm.com/fhir/extension/implicit-system";
+    private static final FHIRGenerator generator = FHIRGenerator.generator(Format.JSON, true);
+
+    public static void main(String[] args) throws Exception {
```
Comment (prb112) [inline fhir-search/src/test/java/com/ibm/fhir/search/util/SearchParameterAugmenter.java:40]: I really like this.

Comment (lmsurpre) [reply fhir-search/src/test/java/com/ibm/fhir/search/util/SearchParameterAugmenter.java:40]: thanks, it was kind of fun to write 👍

**Review by prb112 — APPROVED** [review]

File: fhir-search/src/main/java/com/ibm/fhir/search/util/SearchUtil.java [hunk 2]
```diff
@@ -1253,6 +1254,20 @@ private static void checkSearchParameterCombinations(Class<?> resourceType, List
...
+                    // Optimization for search parameters that always reference the same system, added under #1929
+                    try {
+                        String implicitSystem = searchParameter.getExtension().stream()
+                                .filter(e -> SearchConstants.IMPLICIT_SYSTEM_EXT_URL.equals(e.getUrl()) && e.getValue() != null)
+                                .findFirst()
+                                .map(e -> e.getValue().as(Uri.class).getValue())
+                                .orElse(null);
+                        if (implicitSystem != null) {
+                            parameterValue.setValueSystem(implicitSystem);
+                        }
```
Comment (lmsurpre) [inline fhir-search/src/main/java/com/ibm/fhir/search/util/SearchUtil.java:1266]: this is the meat of the change.  it just takes advantage of the extension that we've added to the SearchParameter definitions (when possible)

**Review by punktilious — APPROVED** [review]
LGTM

**Pull request merged**